README.md
# Advent of Code private stats Part 2 Timer

See the difference in time between gaining 2 stars for the day on your [Advent of Code](https://www.adventofcode.com/) private stats page.

![#Personal Stats](example/aoc-part2-timer-screenshot.png "Part 2 Timer Example")
![#Personal Stats - with puzzle open tracker](example/advent-of-code-part2-timer-tracker.png "Part 2 Timer Example - with puzzle open tracker")


## Installation

* Firefox - https://addons.mozilla.org/en-US/firefox/addon/advent-of-code-part-2-timer/
* Chrome - https://chrome.google.com/webstore/detail/advent-of-code-part-2-timer/fhmjpoppaplfhgnknpbaaklgdnnimfbn

**Note:** you must generate a private leaderboard for this extension to work.


## Building locally

To build the extension locally, run `npm install && npm run build`. This will generate an exploded version of the extension in the `dist/` folder.

Use [`web-ext run`](https://github.com/mozilla/web-ext) from the `dist/` folder to run it locally in Firefox, or `web-ext build` to create a package.


CHANGELOG.md

# Change Log
All notable changes to this project will be documented in this file.
 
The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).

## [1.0.0] - 2021-01-02
### Added
- [Live timer for unfinished Part 2](issues/2)

## [0.0.3] - 2021-01-02
### Added
- [Tracking of "puzzle opens" which allows for offsetting completion time](issues/1)

## [0.0.2] - 2020-12-31  
### Changed
- No longer require manual entry of member ID

## [0.0.1] - 2020-12-30
### Added
- Basic functionality

src/calculate.js
var browser = require("webextension-polyfill");
export function formatTimestamp(seconds, asString = false) {
    var days = 0, hours = 0, minutes = 0
    if (seconds >= 86400) {
        days = Math.floor(seconds / 86400);
        seconds %= 86400;
    }
    if (seconds >= 3600) {
        hours = Math.floor(seconds / 3600);
        seconds %= 3600;
    }
    if (seconds >= 60) {
        minutes = Math.floor(seconds / 60);
        seconds %= 60;
    }
    
    if (asString) {
        return `${hours.toString().padStart(2, 0)}:${minutes.toString().padStart(2, 0)}:${seconds.toString().padStart(2, 0)}`;
    }

    return {
        days,
        hours,
        minutes,
        seconds
    };
}
export function calculateDeltas(data, yearCache) {
    const {owner_id, event, options} = data;
    return new Promise((resolve, reject) => {
        const members = Object.values(data.members);
        const member = members.find(member => member.id == owner_id);
        const completion_day_level = member.completion_day_level
        Object.keys(completion_day_level).forEach(dayNum => {
            const puzzleDropDate = new Date(event, 11, dayNum);
            const obj = completion_day_level[dayNum]
            if (yearCache[dayNum] && yearCache[dayNum] < obj["1"].get_star_ts) {
                const puzzleOpenDate = new Date(yearCache[dayNum] * 1000)
                const daysOffset = Math.floor((puzzleOpenDate - puzzleDropDate) / 86400000)
                obj.puzzleOpen = {
                    ts: yearCache[dayNum],
                    daysOffset,
                    adjustedTS: {
                        1: formatTimestamp(obj["1"].get_star_ts - yearCache[dayNum], true)
                    }
                }
            }
            if (obj["2"]) {
                if (obj.puzzleOpen) {
                    // TODO: don't override if >24 hrs
                    obj.puzzleOpen.adjustedTS["2"] = formatTimestamp(obj["2"].get_star_ts - obj.puzzleOpen.ts, true)
                }
                var delta = obj["2"].get_star_ts - obj["1"].get_star_ts;
                obj.delta = formatTimestamp(delta);
            }
        })
        resolve({completion_day_level, options});
    });
}


src/visualize.js
import { formatTimestamp } from "./calculate";
const lineRegex = new RegExp(/^\s+([^\s]+)(\s+\d+\s+\d+)\s+([^\s]+)(.*)$/);
const origDisplay = document.getElementsByTagName('pre')[0].innerHTML.split("\n");
let vizTimer = NaN

export function visualize(data, options) {
    const includePuzzleOpenedColumn = Object.values(data).some(obj => obj.puzzleOpen != undefined);
    const showTimer = options.showTimer && Object.values(data).some(obj => obj["2"] == undefined);
    let headerStart = origDisplay[0].indexOf("<");
    let updatedDisplay = origDisplay[0].substr(0, headerStart) + (includePuzzleOpenedColumn ? '<span class="leaderboard-daydesc-first">----Puzzle Opened----</span>   ' : '') +
                         origDisplay[0].substr(headerStart) + '   <span class="leaderboard-daydesc-first">---Part 1 to Part 2---</span>\n';
    headerStart = origDisplay[1].indexOf("<");
    updatedDisplay += origDisplay[1].substr(0, headerStart) + ' '.repeat(includePuzzleOpenedColumn ? 24 : 0) + origDisplay[1].substr(headerStart) + "\n";

    origDisplay.slice(2).forEach(line => {
        if (!line.trim()) {
            updatedDisplay += '\n';
            return;
        }
        let [dayNum, restOfLine] = /(\d+)(.*)/.exec(line.trim()).slice(1);
        let delta = "";
        let deltaObj;
        if (data[dayNum].delta) {
            deltaObj = data[dayNum].delta
        } else if (showTimer) {
            const now = Math.floor((new Date()).getTime() / 1000);
            deltaObj = formatTimestamp(now - data[dayNum]["1"].get_star_ts);
        }
        if (deltaObj) {
            delta = `   ${deltaObj.hours.toString().padStart(2, "0")}:${deltaObj.minutes.toString().padStart(2, "0")}:${deltaObj.seconds.toString().padStart(2, "0")}`;
            if (deltaObj.days) {
                delta += ` +${deltaObj.days} days`;
            }
            if (!deltaObj.days && !deltaObj.hours && deltaObj.minutes < 5) {
                delta = '<span class="leaderboard-daydesc-both">' + delta + '</span>';
            } else if (showTimer && !data[dayNum].delta) {
                delta = '<span class="leaderboard-daydesc-first">'  + delta + '</span>';
            }
        }
        let puzzleOpen = ''
        if (data[dayNum].puzzleOpen) {
            const {ts, daysOffset, adjustedTS} = data[dayNum].puzzleOpen;
            const puzzleOpenDate = new Date(ts * 1000);
            if (daysOffset) {
                puzzleOpen = `+${daysOffset} days `;
            }
            puzzleOpen += `${puzzleOpenDate.getHours().toString().padStart(2, 0)}:${puzzleOpenDate.getMinutes().toString().padStart(2, 0)}:${puzzleOpenDate.getSeconds().toString().padStart(2, 0)}`
            let [p1Time, p1RankScore, p2Time, p2RankScore] = lineRegex.exec(restOfLine).slice(1)
            p1Time = adjustedTS["1"]
            if (adjustedTS["2"]) {
                p2Time = adjustedTS["2"]
            }
            puzzleOpen = puzzleOpen.padStart(24, ' ')
            restOfLine = `${p1Time.padStart(11, ' ')}${p1RankScore}  ${p2Time.padStart(9, ' ')}${p2RankScore}`
        }
        if (includePuzzleOpenedColumn) {
            puzzleOpen = puzzleOpen.padStart(24, ' ');   
        }
        updatedDisplay += dayNum.padStart(3, ' ') + puzzleOpen + restOfLine + delta + '\n';
    });
    
    document.getElementsByTagName('pre')[0].innerHTML = updatedDisplay;
    if (!isNaN(vizTimer)) {
        clearTimeout(vizTimer);
    }
    if (showTimer) {
        vizTimer = setTimeout(visualize, 1000, data, options);
    }
    return data;
}

webpack.config.js
const CopyWebpackPlugin = require("copy-webpack-plugin");
const { CleanWebpackPlugin } = require("clean-webpack-plugin");

const package = require("./package.json");

module.exports = {
    mode: "development",
    devtool: "cheap-module-source-map",
    plugins: [
        new CleanWebpackPlugin(),
        new CopyWebpackPlugin([
            { from: "src/addon/icons", to: "icons" },
            { from: "src/options.html", to: "options.html" },
            {
                from: "src/addon/manifest.json",
                to: "manifest.json",
                transform(content, path) {
                    var manifest = JSON.parse(content.toString());
                    manifest.version = package.version;
                    manifestJson = JSON.stringify(manifest, null, 4);
                    return manifestJson;
                }
            }
        ])
    ],
    entry: {
        stats: "./src/stats.js",
        tracker: "./src/tracker.js",
        options: "./src/options.js"
    },
    output: {
        filename: "[name].js"
    }
};


src/loader.js
const browser = require("webextension-polyfill");

function fetchData(url, key) {
    return fetch(url)
        .then(response => response.json())
        .then(data => {
            var val = {
                data: data,
                date: new Date().getTime()
            };
            browser.storage.local.set({ [key]: val });
            return data;
        })
        .catch(err => {
            console.log(err);
        });
}

function getOptions() {
    return new Promise((resolve, reject) => {
        browser.storage.local.get("options").then(options => {
            resolve(options.options ? options.options : { showLiveTimer: false });
        })
    });
}

function getMemberId() {
    return new Promise((resolve, reject) => {
        browser.storage.local.get("memberId").then(memberId => {
            if (memberId === null ||
                !memberId.hasOwnProperty("memberId") ||
                memberId.memberId == null) {

                const settingsUrl = document.location.protocol +
                    "//" +
                    document.location.host +
                    document.location.pathname.split("/").slice(0, 2).join("/") +
                    "/settings";
                const utf8Decoder = new TextDecoder("utf-8");
                fetch(settingsUrl)
                    .then(response => response.body)
                    .then(body => {
                        const reader = body.getReader()
                        reader.read().then(function processText({ done, value }) {
                            if (done) {
                                return;
                            }
                            const memberRegex = new RegExp(/\(anonymous user #(\d+)\)/, 'g');
                            const chunk = utf8Decoder.decode(value);
                            if (chunk.match(memberRegex)) {
                                memberId = memberRegex.exec(chunk)[1]
                                browser.storage.local.set({ memberId });
                                resolve(memberId);
                                return;
                            }

                            return reader.read().then(processText);
                        });
                    })
            } else {
                resolve(memberId.memberId);
            }
        });
    });
}

function getPuzzleOpensForYear(year) {
    return new Promise((resolve, reject) => {
        browser.storage.local.get(year).then(yearCache => {
            resolve(yearCache.hasOwnProperty(year) ? yearCache[year] : {})
        });
    });
}

export function loadData() {
    return new Promise((resolve, reject) => {
        const year = document.location.pathname.split("/")[1];
        getOptions().then(options => {
            getPuzzleOpensForYear(year).then(yearCache => {
                getMemberId().then(memberId => {
                    const url =
                        document.location.protocol +
                        "//" +
                        document.location.host +
                        document.location.pathname.split("/").slice(0, 3).join("/") +
                        "/private/view/" +
                        memberId +
                        ".json";

                    browser.storage.local.get(url).then(cache => {
                        if (
                            cache === null ||
                            !cache.hasOwnProperty(url) ||
                            !cache[url].hasOwnProperty("data") ||
                            !cache[[url]].hasOwnProperty("date")
                        ) {
                            fetchData(url, url).then(data => {
                                data.options = options;
                                resolve({ data, yearCache });
                            });
                        } else {
                            const ttl = new Date(cache[url].date + 5 * 60 * 1000);
                            if (ttl < new Date()) {
                                fetchData(url, url).then(data => {
                                    data.options = options;
                                    resolve({ data, yearCache });
                                });
                            } else {
                                cache[url].data.options = options;
                                resolve({ data: cache[url].data, yearCache });
                            }
                        }
                    });
                });
            });
        });
    });
}
